### PR TITLE
Merge develop into master

### PR DESCRIPTION
**Pull request by joykare**

### CODE REVIEW

File: .env [hunk 1]
```diff
@@ -0,0 +1 @@
+TEST_MONGO_URL=mongodb://localhost/cp3
```
Comment (jkithome) [inline .env]: .env file should not be pushed to the repository.

File: server/controllers/user_controller.js [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+    User.find(function (err, users) {
+      if (err) {
+        res.status(400).send({ message: 'Error occured while accessing the user' });
+      } else {
+        res.status(200).json(users);
+      }
+    });
+  },
+  update: function (req, res) {
+    User.findById(req.params.user_id, function (err, user) {
```
Comment (jkithome) [inline server/controllers/user_controller.js]: While this works, consider using findByIdAndUpdate as it is better.

File: server/controllers/document_controller.js [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+    var skip = req.query.skip || req.headers['skip'];
+    var date = req.query.date;
+    var role = req.query.role;
+
+    if (date) {
+      var start = date + 'T00:00:00Z';
+      var end = date + 'T23:59:59Z';
+
+      Document.find({
+        $and : [ {createdAt: {$gte: start, $lt: end}}, {
```
Comment (jkithome) [inline server/controllers/document_controller.js]: Should be $lte because 23:59:59 is still part of the day.

File: server/controllers/document_controller.js [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+
+        if (documents.length === 0) {
+          res.status(409).send({ message: 'No documents for this date' });
+        }
+        else {
+          res.status(200).json(documents);
+        }
+      });
+    } else if (role) {
+      Role.findOne( {title: role}, function(err, role) {
```
Comment (jkithome) [inline server/controllers/document_controller.js:58]: This would fail due to difference in case, so you use regex i.e  $query.role = new RegExp(role, 'gi');

File: server/controllers/document_controller.js [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+            res.send(documents);
+          });
+      });
+
+    } else {
+      Document.find({
+        $or : [ {ownerId: req.decoded._id}, {accessLevel: 'public'} ]
+      })
+        .skip(parseInt(skip) || 0)
+        .limit(parseInt(limit) || 10)
```
Comment (jkithome) [inline server/controllers/document_controller.js]: If there is no limit it should return all the documents.

File: server/controllers/document_controller.js [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+      var end = date + 'T23:59:59Z';
+
+      Document.find({
+        $and : [ {createdAt: {$gte: start, $lt: end}}, {
+          $or: [
+          {accessLevel: 'public'}, {ownerId: req.decoded._id} ]}
+        ]
+      })
+      .skip(parseInt(skip) || 0)
+      .limit(parseInt(limit) || 10)
```
Comment (jkithome) [inline server/controllers/document_controller.js]: If there is no limit it should return all the documents.

File: server/controllers/user_controller.js [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+            res.status(400).send({ message: 'Error occured while saving the user.' });
+          } else {
+            res.status(200).send({ message: 'User has been updated' });
+          }
+        });
+      }
+    });
+  },
+  find: function (req, res) {
+    User.findOne( {_id: req.params.user_id}, function (err, user) {
```
Comment (jkithome) [inline server/controllers/user_controller.js:131]: Use findById for this.

File: server/controllers/document_controller.js [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+          }
+          else {
+            res.status(200).json(documents);
+          }
+        });
+    }
+  },
+  find: function (req, res) {
+    Document.findById(req.params.document_id)
+      .where('ownerId').equals(req.decoded._id)
```
Comment (jkithome) [inline server/controllers/document_controller.js]: Should find a document where the user seeking it has access i.e document is public or the user is the owner. To clean this up, use a query with all the fields.

File: server/controllers/role_controller.js [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+    Role.find(function (err, roles) {
+      if (err) {
+        res.status(400).send({ message: 'Error occured during request' });
+      } else {
+        res.json(roles);
+      }
+    });
+  },
+  create: function (req, res) {
+    var possibleRoles = Role.schema.path('title').enumValues;
```
Comment (jkithome) [inline server/controllers/role_controller.js]: You don't need to do this. If the role is not allowed it will return an error. You can then parse the error and return the appropriate message.

File: server/models/role.js [hunk 10]
```diff
@@ -0,0 +1,14 @@
+var mongoose = require('mongoose');
+var Schema = mongoose.Schema;
+
+var roleSchema = new Schema ({
+  title: {
+    type: String,
+    required: true,
+    enum: ['superadmin', 'admin', 'user']
+  }
+});
```
Comment (jkithome) [inline server/models/role.js:15]: Your roles need a description field.

File: server/routes/index.js [hunk 11]
```diff
@@ -0,0 +1,65 @@
+var userRoute = require('./user_route'),
+  User = require('../models/user'),
+  documentRoute = require('./document_route'),
+  roleRoute = require('./role_route'),
+  jwt = require('jsonwebtoken'),
+  config = require('../../config/config');
+
+module.exports = function (router) {
+  router.post('/users/login', function (req, res) {
```
Comment (jkithome) [inline server/routes/index.js]: The user login method should be in the users controller. Avoid having logic in your routes files.

File: tests/user.js [hunk 12]
```diff
@@ -0,0 +1,193 @@
...
+  describe('/users CRUD operations', function () {
+    it('route that returns all users', function (done) {
+      request
+        .get('/api/users')
+        .set('x-access-token', token)
+        .expect(200)
+        .end(function (err, res) {
+          expect(res.body).to.have.length(4);
+          expect(Array.isArray(res.body)).to.equal(true);
+          userId = res.body[0]._id;
```
Comment (jkithome) [inline tests/user.js:59]: Assert that array elements are objects like so  expect(Object.prototype.toString.call(res.body[0])).toBe('[object Object]');

File: tests/user.js [hunk 13]
```diff
@@ -0,0 +1,193 @@
...
+          done();
+        });
+    });
+
+    it('creates a user in the db', function (done) {
+      request
+        .post('/api/users')
+        .set('x-access-token', token)
```
Comment (jkithome) [inline tests/user.js]: User creation route should not be authenticated.

File: tests/user.js [hunk 14]
```diff
@@ -0,0 +1,193 @@
...
+        .send({
+          username: 'test',
+          email: 'test@gmail.com',
+          password: 'test',
+          role: 'user'
+        })
+        .expect(400)
+        .expect({message: 'Error occured while saving the user'}, done);
```
Comment (jkithome) [inline tests/user.js:98]: Ideally, if there is a missing value, that should be communicated as opposed to just communicating there was an error.

File: tests/user.js [hunk 15]
```diff
@@ -0,0 +1,193 @@
...
+    });
+
+    it('asserts that a new user has a role defined', function (done) {
+      var role = User.schema.paths.role;
+      expect(role.options.required).to.equal(true);
```
Comment (jkithome) [inline tests/user.js:103]: Ideally, consider returning the user created so that you can confirm the user has a role defined. Minus sensitive details of course. 👍

File: tests/user.js [hunk 16]
```diff
@@ -0,0 +1,193 @@
...
+    });
+
+    it('deletes a particular user inforation', function (done) {
+      request
+        .delete('/api/users/' + userId)
+        .set('x-access-token', token)
+        .expect({message: 'User has been deleted'}, done);
+    });
+  });
+});
```
Comment (jkithome) [inline tests/user.js:199]: You should add a test to confirm that someone cannot edit another person's account as implemented in the code.

File: tests/document.js [hunk 17]
```diff
@@ -0,0 +1,212 @@
...
+            expect(res.body).to.exist;
+            expect(Array.isArray(res.body)).to.equal(true);
+            expect(res.body).to.have.length(2);
+            expect(res.body[0].title).to.equal('User');
+            done();
+          });
+    });
+
+    it('gets number of documents specified published on a certain date', function (done) {
+      var today = new Date();
```
Comment (jkithome) [inline tests/document.js]: Use moment.js to construct the date. It will be a lot easier.

File: tests/document.js [hunk 18]
```diff
@@ -0,0 +1,212 @@
...
+
+      request
+          .get('/api/documents?date=' + date)
+          .set('x-access-token', token)
+          .set('limit', 2)
+          .end(function (err, res) {
+            expect(res.status).to.equal(200);
+            expect(res.body).to.exist;
+            expect(Array.isArray(res.body)).to.equal(true);
+            expect(res.body).to.have.length(2);
```
Comment (jkithome) [inline tests/document.js:134]: To be more accurate with this test, you can seed your documents by inserting them into the database with different created dates.

File: tests/document.js [hunk 19]
```diff
@@ -0,0 +1,212 @@
...
+    });
+
+    it('returns message if no documents are found', function (done) {
+      request
+          .get('/api/documents?date=2016-09-13')
+          .set('x-access-token', token)
+          .expect({message: 'No documents for this date'}, done);
+    });
+
+    it('gets documents published by the same role level(admin)', function (done) {
```
Comment (jkithome) [inline tests/document.js:146]: You should test whether the documents returned are those that are published by a role and accessible to the user.

File: tests/document.js [hunk 20]
```diff
@@ -0,0 +1,212 @@
...
+  describe('/documents/:id test suite', function () {
+    it('returns specific document', function (done) {
+      request
+        .get('/api/documents/' + documentId)
+        .set('x-access-token', token)
+        .end(function (err, res) {
+          expect(res.status).to.equal(200);
+          expect(res.body).to.exist;
+          expect(res.body).to.be.an('object');
+          expect(res.body).to.have.keys('_id', 'title', 'content', 'role', 'modifiedAt', 'ownerId', 'accessLevel', 'createdAt', '__v');
```
Comment (jkithome) [inline tests/document.js]: This could be any document. Confirm with a unique value such as the value of the title field.

File: tests/document.js [hunk 21]
```diff
@@ -0,0 +1,212 @@
...
+        .set('x-access-token', token)
+        .send({
+          title: 'Updated Doc',
+          content: 'Whatever',
+          accessLevel: 'private'
+        })
+        .expect(200)
+        .expect({message: 'Document has been updated'}, done);
+    });
+
```
Comment (jkithome) [inline tests/document.js:200]: Add a test to check whether a person who does not own the document can edit or delete it.

File: tests/role.js [hunk 22]
```diff
@@ -0,0 +1,59 @@
...
+      .set('x-access-token', token)
+      .send({
+        role: 'superadmin'
+      })
+      .end(function (err, res) {
+        expect(res.body.message).to.equal('Role saved');
+        done();
+      });
+  });
+
```
Comment (jkithome) [inline tests/role.js:133]: Add test for creating duplicate role.

File: server/models/user.js [hunk 23]
```diff
@@ -0,0 +1,70 @@
...
+      return next(err);
+    user.password = hash;
+    next();
+  });
+});
+
+// method to compare a give password with the db hash
+userSchema.methods.comparePassword = function (password) {
+  var user = this;
+  return bcrypt.compareSync(password, user.password);
```
Comment (andela-kndungu) [inline server/models/user.js]: *Code Quality - Moderate*

It's better to use asynchronous functions in a node environment.

Comment (kndungu) [inline server/models/user.js:67]: __Code Quality - Critical__

Better to use asynchronous function equivalient

File: server/controllers/user_controller.js [hunk 24]
```diff
@@ -0,0 +1,196 @@
+var User = require('../models/user.js');
```
Comment (kndungu) [inline server/controllers/user_controller.js:1]: __Nitpick__

Use camel case when naming your files.

File: server/controllers/user_controller.js [hunk 25]
```diff
@@ -0,0 +1,196 @@
...
+var config = require('../../config/config');
+
+module.exports = {
+  login: function (req, res) {
+    User.findOne( {email: req.body.email} )
+      .select('email password role')
+      .populate('role')
+      .exec(function (err, user) {
+        if (err) {
+          res.status(400).send({ message: 'Failed to log in, try again' });
```
Comment (kndungu) [inline server/controllers/user_controller.js:14]: __Code Quality - Moderate__

Being a server error the status code should be of the 5*\* family

File: server/controllers/user_controller.js [hunk 26]
```diff
@@ -0,0 +1,196 @@
...
+        }
+        if (!user) {
+          res.status(401).send({ message: 'Email does not exist' });
```
Comment (kndungu) [inline server/controllers/user_controller.js:17]: __Code Quality - Moderate__

401 is for unauthorised, 404 is more applicable in this case

File: server/controllers/user_controller.js [hunk 27]
```diff
@@ -0,0 +1,196 @@
+var User = require('../models/user.js');
+var Document = require('../models/document.js');
+var Role = require('../models/role.js');
+var jwt = require('jsonwebtoken');
+var config = require('../../config/config');
+
+module.exports = {
+  login: function (req, res) {
```
Comment (kndungu) [inline server/controllers/user_controller.js:8]: __Best Practice__

In all your functions add return when sending a response,

File: config/config.js [hunk 28]
```diff
@@ -0,0 +1,7 @@
+require('dotenv').config({silent:true});
+
+module.exports = {
+  test_database: process.env.TEST_MONGO_URL,
+  port: process.env.PORT || 8080,
+  secret: 'whatwhatwhatwhatwhat'
```
Comment (kndungu) [inline config/config.js:6]: Sensitive information, have it in the .env file

File: server/controllers/document_controller.js [hunk 29]
```diff
@@ -0,0 +1,156 @@
...
+
+    document.save(function (err) {
+      if (err) {
+        if (err.code === 11000) {
+          res.status(409).send({ message: 'Duplicate entry' });
+        } else {
+          res.status(400).send({ message: 'Error occured while saving the document' });
+        }
+      } else {
+        res.status(200).send({ message: 'New document created' });
```
Comment (kndungu) [inline server/controllers/document_controller.js:22]: __Best Practice__

Return the newly created document

File: server/controllers/document_controller.js [hunk 30]
```diff
@@ -0,0 +1,156 @@
...
+  },
+  all: function (req, res) {
+    var limit = req.query.limit || req.headers['limit'];
+    var skip = req.query.skip || req.headers['skip'];
+    var date = req.query.date;
+    var role = req.query.role;
+
+    if (date) {
+      var start = date + 'T00:00:00Z';
+      var end = date + 'T23:59:59Z';
```
Comment (kndungu) [inline server/controllers/document_controller.js:34]: You should not treat dates as strings, Mongo can query date objects

**Review by kndungu — COMMENTED** [review]
Kinuthia Review
